### ISSUE TITLE
Reassign Future Approver does not work on Alfresco

### ISSUE THREAD

**Issue by gsteimer** [tsgrp/HPI#1967]
Documentum - when the future task has more than one approver. After reassigning a future approver and finishing outstanding items to get to that task:
- [x] , only the reassigned user actually gets the task.  Additionally, when the reassigned approver approves, no signature is added to the signature page.

In the following example, I have two tasks IQA (qdadmin3) and CR Approval (cobrien, mblum)

Pre route:
![image](https://user-images.githubusercontent.com/1377228/66416584-c258ba00-e9c3-11e9-9aa1-a63ffd62bcc1.png)

Started route, reassigned mblum to qdauthor1
![image](https://user-images.githubusercontent.com/1377228/66416626-d7cde400-e9c3-11e9-82fd-eb2d152361cc.png)

Approved as qdadmin3.  QD Author1 inbox:
![image](https://user-images.githubusercontent.com/1377228/66416725-1ebbd980-e9c4-11e9-8911-755db16c01ab.png)

Conan's Inbox - doesn't have the task:
![image](https://user-images.githubusercontent.com/1377228/66416819-5460c280-e9c4-11e9-94ca-5696320acdb8.png)

After approving as qdauthor1 - no sig page
![image](https://user-images.githubusercontent.com/1377228/66416776-41e68900-e9c4-11e9-9ffc-7548d962755d.png)

Trying to cancel workflow as original user - interface showing "Group Assignee"
![image](https://user-images.githubusercontent.com/1377228/66417037-c507df00-e9c4-11e9-9237-7a312d5185c7.png)

Trying to reassign the task - it still thinks that Conan has a task
![image](https://user-images.githubusercontent.com/1377228/66417083-da7d0900-e9c4-11e9-95f2-ab5f8207d8c4.png)

The above must be fixed for project Jurrasic Park.  After those issues are fixed, I want to retest some oddities with reassigning a future task that I only see on Alfresco:

- [x] After reassigning a future approver:
  - After approving the current task, the approver never shows as Approved in the workflow status
  - The signature isn't applied to the form properly for this current task
- [x] When approving as the reassigned approver, the signature page text isn't finding the original approver.  The text is: `... on behalf of null`

**Comment by mwojno-tsg** [comment]
The issue is resolved in OC commit 23506. The changes that were made was updating the reassign executer's logic to support the ability to support multiple role future reassignments, the logic was also solidified to make sure it is correctly setting and giving users their correct tasks. Circular reassignment list was added as well to safeguard against a user ending up with the task they started with in future reassignments. Finally, the signature was updated to make sure the original user information was set so the signature page and canceling a form should work now.

**Comment by gsteimer** [comment]
Reassign and Delegate are failing on Alfresco.  When you search for a user to delegate (or reassign) the task to, the query fails in OC with the following error:

```
2019-10-21 16:09:58,125	ajp-apr-8009-exec-10	ERROR	RESTService	Unhandled exception around OC method class: interface com.tsgrp.opencontent.core.user.IUser| name: searchAuthorities
com.tsgrp.opencontent.core.exception.OCRuntimeException: Unhandled exception around OC method class: interface com.tsgrp.opencontent.core.user.IUser| name: searchAuthorities
	at com.tsgrp.opencontent.alfresco.core.EmbeddedAlfrescoAdvice.invoke(EmbeddedAlfrescoAdvice.java:98) ~[oc-hpiDemoAlf.2.alfrescoEmb-3.2.0.jar:3.2.0]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[spring-aop-3.2.14.RELEASE.jar:3.2.14.RELEASE]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204) ~[spring-aop-3.2.14.RELEASE.jar:3.2.14.RELEASE]
	at com.sun.proxy.$Proxy462.searchAuthorities(Unknown Source) ~[?:?]
	at com.tsgrp.opencontent.core.user.RestUser.getSearchAuthorities(RestUser.java:511) ~[oc-hpiDemoAlf.1.core-3.2.0.jar:3.2.0]
...
Caused by: java.lang.NullPointerException
	at com.tsgrp.opencontent.alfresco.user.AlfrescoEmbUserImpl.searchAuthorities(AlfrescoEmbUserImpl.java:884) ~[oc-hpiDemoAlf.2.alfrescoEmb-3.2.0.jar:3.2.0]
```

The line it's failing on is:

```java
String groupNodeRef = serviceRegistry.getAuthorityService().getAuthorityNodeRef(criteria.getChildOfAuthority()).toString();
```

**Comment by gsteimer** [comment]
Ok - I think I figured out what's going on.  The switch to the search authorities endpoint generates a URL like this:

`https://edge2.tsgrp.com/alfresco/OpenContent/authorities/search?authorityType=user&childOfAuthority=wizard_contributors&attrToSearch=displayName&matchType=contains&isCaseSensitive=false&query=QD`

... which defaults the group to `wizard_contributors`.  However, in Alfresco the group name is actually `GROUP_wizard_contributors`.

This begs a long standing question - should the alfresco searchAuthorities impl check for `GROUP_` before the group name when running a search, and if it's not there add it?

**Comment by gsteimer** [comment]
Above user search issue fixed with OC REV 23574

**Comment by gsteimer** [comment]
With the latest round of updates, reassigning a future approver on alfresco seems to work fine, but for some reason the _current_ task is not being processed properly.  For example,

Route start:
![image](https://user-images.githubusercontent.com/1377228/67310352-36668800-f4c3-11e9-9917-ab0674ce0cb7.png)

Reassign future approver (Alice):
![image](https://user-images.githubusercontent.com/1377228/67310387-45e5d100-f4c3-11e9-99da-971e04b24c63.png)

Approve as IQA (QD Admin 3):
- Signature applied properly
- Task removed from users inbox
_But - Approval Details still lists the user as pending_:
![image](https://user-images.githubusercontent.com/1377228/67310515-82193180-f4c3-11e9-9e71-745ad4836352.png)

Because of this when Mike approves his task, the signature for QD Admin 3 is wiped out.

**Comment by mwojno-tsg** [comment]
Temporary fix was committed on OCMS commit: 23989
A slider (kill switch) was added in the admin console to toggle the future reassignment functionality on or off.

**Comment by mwojno-tsg** [comment]
Thoughts on what is going on
- To summarize: what is going on is a communication issue between the CompleteUserTaskListener and the WizardApprovalWorkflowSubprocessOnAssignment. What happens is when the user goes to approve the task, it successfully goes through the CompleteUserTaskListener. The CompleteUserTaskListener will, among other things, update the approval details to "approved" not "pending". Next, it goes into  WizardApprovalWorkflowSubprocessOnAssignment to do the reassignments. However, the approval details it gets are not the updated one that CompleteUserTaskListener just changed. So, during WizardApprovalWorkflowSubprocessOnAssignment's logic it will change the approval details based on its reassignments it needs to do and will overwrite the approval details with incorrect signature statuses. This, leads to the above issue where the approval details will show it as "pending" not "approved".

- Here is where exactly in code I am talking about and where the overwriting happens in the WizardApprovalWorkflowSubprocessOnAssignment line 173: 
![image](https://user-images.githubusercontent.com/54864662/67595768-f4407f00-f72c-11e9-90bb-e4fc53dc52ea.png)

Thoughts on possible fixes
- An initial fix I thought of was messing with the futureReassignment variable and trying to have checks to control when ^ line 173 is hit. However, this is pretty tricky because that line needs to be hit in a number of occasions. One such occasion is when we reassign a future task, the previousAssigneeLogin and newAssigneeLogin are the same until we hit the conditional of futureReassignment to then change newAssigneeLogin to what it needs to be. This will call the listener again because task.setAssignee() is called. We now need to hit line 173 to update the approval details. This conflicts with the case when we want to safeguard the overwriting on line 173, by checking if it is an active task. These events are so intertwined that I don't think it as simple as safeguarding with conditionals here and could get even more messy than it already is.

- Another solution that would be a little more architecture change is the moving of updating approval details. Right now all the reassign and delegate executors do for their assigning is setting the task assignee on the task id. Then, it is the job of WizardApprovalWorkflowSubprocessOnAssignment to actually do the updating of approval details and more. So, maybe there is a better place to be doing this approval details updating and a little more restructuring of these listener files? I think there would need to be a meeting to discuss this because so of the high level understanding of these listener communications is lacking.

- A somewhat hacky solution could be an addition of another executionVariable that could keep track of if a future reassignment happened. This would be a little tricky and would have to trace the different cases through to make sure all our covered.

Conclusion
- This fix is a little extensive for the miscommunication between both listener and will need a little more thought and testing to figure a better solution that could work on all cases. For now, there is a kill switch to toggle this functionality on or off

**Comment by gsteimer** [comment]
Note - when getting back to this, we should test this on Hbase as well.  The thought was that this would work in a repository agnostic manner, so it _should_ work, but we thought that with Alfresco as well.

**Comment by mwojno-tsg** [comment]
In commit #25176 for the paas project, the reassign future functionality was turned off for alfresco because it correctly does not work as detailed in the above issue. The real issue here is now being tracked in a consolidated issue, see #2107

**Comment by cheinz-tsg** [comment]
Closed with OC Commit #25614, which fixed the root cause (issue #2107). Now that the AppDetails bug is fixed and the most updated version of the object is returned in Alfresco, no further changes were needed to make this functionality work in Alfresco-backed instances.
